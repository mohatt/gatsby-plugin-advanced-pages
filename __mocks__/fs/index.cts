import { fs } from 'memfs'
import __realFs from 'fs'

/**
 * An internal reference to the real Node.js `fs` module.
 * Provides access to the native filesystem instance.
 */
Object.assign(fs, { __realFs })

module.exports = fs
