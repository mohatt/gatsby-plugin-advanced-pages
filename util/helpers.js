const path = require('path')
const debug = require("debug")("gatsby-plugin-advanced-pages")

// Checks if a path is a child of another one
function isPathChildof (child, parent) {
  const relative = path.relative(parent, child)
  return relative && !relative.startsWith('..') && !path.isAbsolute(relative)
}

module.exports = {
  debug,
  isPathChildof,
}
