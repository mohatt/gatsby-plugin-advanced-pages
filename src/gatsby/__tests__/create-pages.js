import fs from 'fs'
import path from 'path'
import { createPages } from '../'
import { mountOptions, mountFile, mountDir, mountModule } from '../../../test/helpers'
import testCases from '../../../test/__fixtures__/create-pages'

// Use in-memory file system
jest.mock('fs')

describe('createPages', () => {
  const graphql = jest.fn()
  const createPage = jest.fn()
  const actions = { createPage }
  const getNodesByType = jest.fn().mockImplementation(() => 10)
  const helperFile = '/path/to/helper.js'

  beforeEach(() => {
    graphql.mockReset()
    createPage.mockReset()
    fs.reset()
    jest.resetModules()
    mountOptions()
  })

  for (const { id, throws, pages, helper } of testCases) {
    // Set test title
    const title = throws
      ? `throws error on (${id})`
      : `correctly creates pages on (${id})`

    test(title, async () => {
      if (helper) {
        mountFile(helperFile)
        mountModule(helperFile, helper)
      }

      // Create a virtual directory for the 'src' folder
      // so that routes.js file can be written virtually
      mountDir(path.resolve(__dirname, '../../'))

      graphql.mockReturnValue({
        data: {
          allPage: { nodes: pages }
        }
      })

      let error = null
      try {
        await createPages({
          graphql,
          actions,
          getNodesByType
        })
      } catch (e) {
        error = e
      }

      if (throws) {
        expect(error).not.toBeNull()
        expect(error).toMatchSnapshot()
        return
      }

      expect(error).toBeNull()
      expect(createPage.mock.calls).toMatchSnapshot()
      expect(
        fs.readFileSync(path.resolve(__dirname, '../../routes.js'), 'utf8')
      ).toMatchSnapshot()
    })
  }
})
