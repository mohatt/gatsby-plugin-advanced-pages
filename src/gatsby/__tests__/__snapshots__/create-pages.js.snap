// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPages correctly creates pages on (blog-paginated-auto-route) 1`] = `
Array [
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 10,
        "offset": 0,
      },
      "path": "/blog",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 10,
        "offset": 10,
      },
      "path": "/blog/page/2",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 10,
        "offset": 20,
      },
      "path": "/blog/page/3",
    },
  ],
]
`;

exports[`createPages correctly creates pages on (blog-paginated-auto-route) 2`] = `"[{\\"name\\":\\"blog\\",\\"path\\":\\"/blog\\",\\"scopes\\":{\\"pagination\\":\\"/blog/page/:page\\"}}]"`;

exports[`createPages correctly creates pages on (blog-paginated-custom-limit) 1`] = `
Array [
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 8,
        "offset": 0,
      },
      "path": "/blog",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 8,
        "offset": 8,
      },
      "path": "/blog/page/2",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 8,
        "offset": 16,
      },
      "path": "/blog/page/3",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 8,
        "offset": 24,
      },
      "path": "/blog/page/4",
    },
  ],
]
`;

exports[`createPages correctly creates pages on (blog-paginated-custom-limit) 2`] = `"[{\\"name\\":\\"blog\\",\\"path\\":\\"/blog\\",\\"scopes\\":{\\"pagination\\":\\"/blog/page/:page\\"}}]"`;

exports[`createPages correctly creates pages on (blog-paginated-manual-route) 1`] = `
Array [
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 10,
        "offset": 0,
      },
      "path": "/blog",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 10,
        "offset": 10,
      },
      "path": "/blog/what/2",
    },
  ],
  Array [
    Object {
      "component": "/path/to/blog.js",
      "context": Object {
        "id": "blog-id",
        "limit": 10,
        "offset": 20,
      },
      "path": "/blog/what/3",
    },
  ],
]
`;

exports[`createPages correctly creates pages on (blog-paginated-manual-route) 2`] = `"[{\\"name\\":\\"blog\\",\\"path\\":\\"/blog\\",\\"scopes\\":{\\"pagination\\":\\"/blog/what/:page\\"}},{\\"name\\":\\"blog.paginated\\",\\"path\\":\\"/blog/what/:page\\",\\"scopes\\":{}}]"`;

exports[`createPages correctly creates pages on (dynamic-page) 1`] = `
Array [
  Array [
    Object {
      "component": "/path/to/page.js",
      "context": Object {
        "id": "page-id",
        "page": "hello",
      },
      "path": "/pages/hello",
    },
  ],
  Array [
    Object {
      "component": "/path/to/page.js",
      "context": Object {
        "id": "page-id",
        "page": "world",
      },
      "path": "/pages/world",
    },
  ],
]
`;

exports[`createPages correctly creates pages on (dynamic-page) 2`] = `"[{\\"name\\":\\"page\\",\\"path\\":\\"/pages/:page\\",\\"scopes\\":{}}]"`;

exports[`createPages correctly creates pages on (home-about) 1`] = `
Array [
  Array [
    Object {
      "component": "/path/to/index.js",
      "context": Object {
        "id": "home-id",
      },
      "path": "/",
    },
  ],
  Array [
    Object {
      "component": "/path/to/page.js",
      "context": Object {
        "id": "about-id",
      },
      "path": "/about",
    },
  ],
  Array [
    Object {
      "component": "/path/to/page.js",
      "context": Object {
        "id": "about-id",
      },
      "path": "/aboutme",
    },
  ],
]
`;

exports[`createPages correctly creates pages on (home-about) 2`] = `"[{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"scopes\\":{}},{\\"name\\":\\"about\\",\\"path\\":\\"/about\\",\\"scopes\\":{}},{\\"name\\":\\"aboutme\\",\\"path\\":\\"/aboutme\\",\\"scopes\\":{}}]"`;

exports[`createPages correctly creates pages on (page-context) 1`] = `
Array [
  Array [
    Object {
      "component": "/path/to/page.js",
      "context": Object {
        "id": "about-id",
        "name": "adam",
        "profile": Object {
          "age": 25,
          "gender": "male",
        },
      },
      "path": "/about/adam",
    },
  ],
  Array [
    Object {
      "component": "/path/to/page.js",
      "context": Object {
        "id": "about-id",
        "name": "sara",
        "profile": Object {
          "age": 28,
          "gender": "female",
        },
      },
      "path": "/about/sara",
    },
  ],
]
`;

exports[`createPages correctly creates pages on (page-context) 2`] = `"[{\\"name\\":\\"about\\",\\"path\\":\\"/about/:name\\",\\"scopes\\":{}}]"`;

exports[`createPages throws error on (bad-helper-empty-call) 1`] = `[Error: Route name passed to createAdvancedPage() at "/path/to/helper.js" must be a non-empty string]`;

exports[`createPages throws error on (bad-helper-empty-pagination) 1`] = `[Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": "count" paramater is missing]`;

exports[`createPages throws error on (bad-helper-invalid-pagination-count) 1`] = `[Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": "count" paramater must be a valid non-negative number (got "-15")]`;

exports[`createPages throws error on (bad-helper-invalid-pagination-limit) 1`] = `[Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": "limit" paramater must be a valid positive number (got "-10")]`;

exports[`createPages throws error on (bad-helper-invalid-pagination-route) 1`] = `[Error: Invalid pagination object passed to createAdvancedPage() at "/path/to/helper.js": Unrecognized route "foo"]`;

exports[`createPages throws error on (bad-helper-invalid-route-params) 1`] = `[TypeError: Error generating a path for route "/about/:name" with params "{}": Expected "name" to be a string]`;

exports[`createPages throws error on (bad-helper-js-error) 1`] = `[Error: some error]`;

exports[`createPages throws error on (bad-helper-undefined-route) 1`] = `[Error: Unrecognized route "blog" passed to createAdvancedPage() at "/path/to/helper.js"]`;

exports[`createPages throws error on (duplicated-routes) 1`] = `[Error: Unable to create a route with name "about". Another route with the same name already exists]`;

exports[`createPages throws error on (invalid-helper) 1`] = `[Error: Cannot find module '/path/to/whatever.js' from 'src/gatsby/lib/pages-creator.js']`;
